dotnet/src/Agents/Abstractions/AgentChat.cs

```diff
@@ -134,47 +193,81 @@ await foreach (var message in channel.InvokeAsync(agent, cancellationToken).Conf
         }
         finally
         {
-            Interlocked.Exchange(ref this._isActive, 0);
+            this.ClearActivitySignal(); // Signal activity hash completed
         }
-    }
 
-    private async Task<AgentChannel> GetChannelAsync(Agent agent, CancellationToken cancellationToken)
-    {
-        var channelKey = this.GetAgentHash(agent);
-
-        if (this._agentChannels.TryGetValue(channelKey, out var channel))
-        {
-            await this._broadcastQueue.EnsureSynchronizedAsync(new ChannelReference(channel, channelKey)).ConfigureAwait(false);
-        }
-        else
+        async Task<AgentChannel> GetOrCreateChannelAsync()
         {
-            channel = await agent.CreateChannelAsync(cancellationToken).ConfigureAwait(false);
-
-            if (this.History.Count > 0)
+            string channelKey = this.GetAgentHash(agent);
+            AgentChannel channel = await this.SynchronizeChannelAsync(channelKey, cancellationToken).ConfigureAwait(false);
+            if (channel == null)
             {
-                await channel.ReceiveAsync(this.History, cancellationToken).ConfigureAwait(false);
+                channel = await agent.CreateChannelAsync(cancellationToken).ConfigureAwait(false);
+                this._agentChannels.Add(channelKey, channel);
+
+                if (this.History.Count > 0)
+                {
+                    await channel.ReceiveAsync(this.History, cancellationToken).ConfigureAwait(false);
+                }
             }
 
-            this._agentChannels.Add(channelKey, channel);
+            return channel;
         }
+    }
 
-        return channel;
+    /// <summary>
+    /// Clear activity signal to indicate that activity has ceased.
+    /// </summary>
+    private void ClearActivitySignal()
+    {
+        // Note: Interlocked is the absolute lightest synchronization mechanism available in dotnet.
+        Interlocked.Exchange(ref this._isActive, 0);
     }
 
-    private string GetAgentHash(Agent agent)
+    /// <summary>
+    /// Test to ensure chat is not concurrently active and throw exception if it is.
+    /// If not, activity is signaled.
+    /// </summary>
+    /// <remarks>
+    /// Rather than allowing concurrent invocation to result in undefined behavior / failure,
+    /// it is preferred to fail-fast in order to avoid side-effects / state mutation.
+    /// The activity signal is used to manage ability and visibility for taking actions based
+    /// on conversation history.
+    /// </remarks>
+    private void ThrowIfActive()
```

Comment:
The name implies checking the state rather than setting it. I would propose to name it in such a way that it's clear it not only checks and throws, but also sets the signal - SetActivityOrThrow?

Reply:
Updated